src/analysis.ts

```diff
@@ -36,10 +39,55 @@ export class ValidationError extends Error {
 
 
 export class Analysis {
-  descriptors: DocumentDescriptor[];
+  private _descriptors: DocumentDescriptor[];
+  private _elementsByTagName = new Map<string, ElementDescriptor>();
+  private _elementsByPackageDir = new Map<string, ElementDescriptor[]>();
+  elementPaths = new Map<ElementDescriptor, string>();
+  locationOffsets = new Map<ElementDescriptor, LocationOffset>();
 
   constructor(descriptors: DocumentDescriptor[]) {
-    this.descriptors = descriptors;
+    this._descriptors = descriptors;
+    const packageGatherer = new PackageGatherer();
+    const elementsGatherer = new ElementGatherer();
+    new AnalysisWalker(this._descriptors).walk([
+      packageGatherer, elementsGatherer
+    ]);
+
+    const elements = elementsGatherer.elementDescriptors;
+
+    for (const element of elements) {
+      if (element.is) {
+        this._elementsByTagName.set(element.is);
+      }
+      let elementPath = elementsGatherer.elementPaths.get(element);
+      const matchingPackageDirs =
+          Array.from(packageGatherer.packageDirs)
+              .filter(dir => trimLeft(elementPath, '/').startsWith(dir));
+      const longestMatchingPackageDir =
+          matchingPackageDirs.sort((a, b) => b.length - a.length)[0] || '';
+
+      const elementsInPackage =
+          this._elementsByPackageDir.get(longestMatchingPackageDir) || [];
+      elementsInPackage.push(element);
+      this._elementsByPackageDir.set(
+          longestMatchingPackageDir, elementsInPackage);
+      // We want element paths to be relative to the package directory.
+      elementsGatherer.elementPaths.set(
+          element, trimLeft(
```

Comment:
why trim operations on paths? is there a `path` operation that does what you want?